secretstores/tencent/ssm/ssm.go

```diff
@@ -0,0 +1,179 @@
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package ssm
+
+import (
+	"context"
+	"encoding/base64"
+	"errors"
+
+	"github.com/dapr/components-contrib/secretstores"
+	"github.com/dapr/kit/logger"
+	"github.com/tencentcloud/tencentcloud-sdk-go/tencentcloud/common"
+	"github.com/tencentcloud/tencentcloud-sdk-go/tencentcloud/common/profile"
+	ssm "github.com/tencentcloud/tencentcloud-sdk-go/tencentcloud/ssm/v20190923"
+)
+
+const (
+	SecretID         = "SecretID"
+	SecretKey        = "SecretKey"
+	Token            = "Token"
+	Region           = "Region"
+	VersionID        = "VersionID"
+	RequestID        = "RequestID"
+	pageLimit uint64 = 100
+)
+
+type ssmClient interface {
+	ListSecretsWithContext(ctx context.Context, request *ssm.ListSecretsRequest) (*ssm.ListSecretsResponse, error)
+	GetSecretValueWithContext(ctx context.Context, request *ssm.GetSecretValueRequest) (*ssm.GetSecretValueResponse, error)
+}
+
+type ssmSecretStore struct {
+	client ssmClient
+	logger logger.Logger
+}
+
+var (
+	_ secretstores.SecretStore = (*ssmSecretStore)(nil)
+)
+
+// NewSSM returns a new TencentCloud ssm secret store.
+func NewSSM(logger logger.Logger) secretstores.SecretStore {
+	return &ssmSecretStore{
+		logger: logger,
+	}
+}
+
+// Init creates a TencentCloud ssm client.
+func (s *ssmSecretStore) Init(metadata secretstores.Metadata) error {
+	var (
+		err    error
+		region string
+	)
+
+	secretID := metadata.Properties[SecretID]
+	secretKey := metadata.Properties[SecretKey]
+	token := metadata.Properties[Token]
+	region = metadata.Properties[Region]
+	if secretID == "" || secretKey == "" {
+		return errors.New("secret params are empty")
+	}
+
+	credential := common.NewTokenCredential(secretID, secretKey, token)
+	s.client, err = ssm.NewClient(credential, region, profile.NewClientProfile())
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// GetSecret retrieves a secret using a key and returns a map of decrypted string/string values.
+func (s *ssmSecretStore) GetSecret(ctx context.Context, req secretstores.GetSecretRequest) (secretstores.GetSecretResponse, error) {
+	response := secretstores.GetSecretResponse{
+		Data: map[string]string{},
+	}
+	if req.Name == "" {
+		return response, errors.New("secret name is empty")
+	}
+
+	versionID := req.Metadata[VersionID]
+	ssmReq := ssm.NewGetSecretValueRequest()
+	ssmReq.SecretName = &req.Name
+	ssmReq.VersionId = &versionID
+	ssResp, err := s.client.GetSecretValueWithContext(ctx, ssmReq)
+	if err != nil {
+		return response, err
+	}
+	var val string
+	ssResponse := ssResp.Response
+	if ssResponse.SecretBinary != nil {
+		bts, err := base64.StdEncoding.DecodeString(*ssResponse.SecretBinary)
```

Comment:
Wont this return a binary string that could be corrupted? Wouldn't it be best to return the secret as base64-encoded here?

For example, imagine the secret was a 256-bit symmetric key... You probably don't want that in a `string` variable. If it has to be a string, it should probably be base64-encoded

Reply:
I originally thought that the upstream micrologic does not need to care about the decoding problem. After ensuring that `SecretString` and `SecretBinary` are aligned, they are returned to the micrologic server, so that when the micrologic server gets the value, they don't need to care what type it is. If we want to transfer primitive values here, we may also need to tell the type of values.

Reply:
Doing so, it may be a little burden.

Reply:
It depends on what the value inside `SecretBinary` is. If it's just an ASCII (or at most, UTF-8) encoded string, then decoding the base64 and converting it to `string` should be fine.

BUT, if the value is a binary string (i.e something that could contain byte sequences that are NOT valid UTF-8 sequences), then converting to `string` could break the content.

Reply:
> Wont this return a binary string that could be corrupted? Wouldn't it be best to return the secret as base64-encoded here?
> 
> For example, imagine the secret was a 256-bit symmetric key... You probably don't want that in a `string` variable. If it has to be a string, it should probably be base64-encoded

After base64 decoding, it is transmitted through the body, is it possible to cause damage?

Reply:
Yes. Just converting it to `string` could cause damage. If it's binary, please leave it in a `[]byte` or leave it as a base64-encoded string.

Reply:
I added a SecretValueType Field in return map to classfiy text and binary data.

Reply:
> Yes. Just converting it to `string` could cause damage. If it's binary, please leave it in a `[]byte` or leave it as a base64-encoded string.

Ok, I see it. Thanks.